Regression: rejected client connection decrement pool counter

After fix for #285, spring-messaging has observed a failing test scenario with the following stack:
```
2018-02-18 14:30:27,578 WARN [tcp-client-loop-nio-1] [io.netty.util.concurrent.DefaultPromise] - An exception was thrown by io.netty.channel.pool.FixedChannelPool$4.operationComplete()
java.lang.AssertionError: null
	at io.netty.channel.pool.FixedChannelPool.decrementAndRunTaskQueue(FixedChannelPool.java:324) ~[netty-transport-4.1.21.Final.jar:4.1.21.Final]
	at io.netty.channel.pool.FixedChannelPool.access$600(c.java:38) ~[netty-transport-4.1.21.Final.jar:4.1.21.Final]
	at io.netty.channel.pool.FixedChannelPool$4.operationComplete(FixedChannelPool.java:305) ~[netty-transport-4.1.21.Final.jar:4.1.21.Final]
```

The stack points to an illegal counting occurring in `FixedChannelPool`. The scenario involves an immediately rejected connection to reproduce.

--- a/src/main/java/reactor/ipc/netty/channel/PooledClientContextHandler.java
+++ b/src/main/java/reactor/ipc/netty/channel/PooledClientContextHandler.java
@@ -200,6 +200,9 @@ final class PooledClientContextHandler<CHANNEL extends Channel>
 		if (createOperations(c, null) == null) {
 			setFuture(pool.acquire());
 		}
+		else {
+			c.closeFuture().addListener(ff -> pool.release(c));
+		}
 	}
 
 	@Override

--- a/src/main/java/reactor/ipc/netty/channel/PooledClientContextHandler.java
+++ b/src/main/java/reactor/ipc/netty/channel/PooledClientContextHandler.java
@@ -189,7 +189,6 @@ final class PooledClientContextHandler<CHANNEL extends Channel>
 		if (!c.isActive()) {
 			log.debug("Immediately aborted pooled channel, re-acquiring new " + "channel: {}",
 					c.toString());
-			release(c);
 			setFuture(pool.acquire());
 			return;
 		}
@@ -200,10 +199,9 @@ final class PooledClientContextHandler<CHANNEL extends Channel>
 		if (createOperations(c, null) == null) {
 			setFuture(pool.acquire());
 		}
-		else {
-			c.closeFuture().addListener(ff -> pool.release(c));
-		}
 	}
+
+
 
 	@Override
 	@SuppressWarnings("unchecked")

--- a/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
+++ b/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
@@ -161,6 +161,7 @@ final class DefaultPoolResources implements PoolResources {
 
 		@Override
 		public void channelCreated(Channel ch) throws Exception {
+			activeConnections.incrementAndGet();
 			/*
 				Sometimes the Channel can be notified as created (by FixedChannelPool) but
 				it actually fails to connect and the FixedChannelPool will decrement its
@@ -172,6 +173,7 @@ final class DefaultPoolResources implements PoolResources {
 
 				see https://github.com/reactor/reactor-netty/issues/289
 			 */
+
 			if (log.isDebugEnabled()) {
 				log.debug("Created new pooled channel {}, now {} active connections",
 						ch.toString(),

--- a/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
+++ b/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
@@ -157,8 +157,6 @@ final class DefaultPoolResources implements PoolResources {
 						ch.toString(),
 						activeConnections);
 			}
-			ch.closeFuture()
-			  .addListener(ff -> pool.release(ch).get());
 		}
 
 		@Override
@@ -170,12 +168,12 @@ final class DefaultPoolResources implements PoolResources {
 				still invoked, which can lead to double-decrement and an assertion error.
 
 				As such, it is best to only register the close handler on the channel in
-				`channelAcquired`.
+				`PooledClientContextHandler`.
 
 				see https://github.com/reactor/reactor-netty/issues/289
 			 */
 			if (log.isDebugEnabled()) {
-				log.debug("Created new pooled channel {}, pending channelAcquired, had {} active connections",
+				log.debug("Created new pooled channel {}, now {} active connections",
 						ch.toString(),
 						activeConnections);
 			}

--- a/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
+++ b/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
@@ -157,21 +157,28 @@ final class DefaultPoolResources implements PoolResources {
 						ch.toString(),
 						activeConnections);
 			}
+			ch.closeFuture()
+			  .addListener(ff -> pool.release(ch).get());
 		}
 
 		@Override
 		public void channelCreated(Channel ch) throws Exception {
-			activeConnections.incrementAndGet();
+			/*
+				Sometimes the Channel can be notified as created (by FixedChannelPool) but
+				it actually fails to connect and the FixedChannelPool will decrement its
+				active count, same as if it was released. The channel close promise is
+				still invoked, which can lead to double-decrement and an assertion error.
+
+				As such, it is best to only register the close handler on the channel in
+				`channelAcquired`.
+
+				see https://github.com/reactor/reactor-netty/issues/289
+			 */
 			if (log.isDebugEnabled()) {
-				log.debug("Created {}, now {} active connections",
+				log.debug("Created new pooled channel {}, pending channelAcquired, had {} active connections",
 						ch.toString(),
 						activeConnections);
 			}
-			if (log.isDebugEnabled()) {
-				log.debug("Created new pooled channel: " + ch.toString());
-			}
-			ch.closeFuture()
-			  .addListener(ff -> pool.release(ch));
 			if (onChannelCreate != null) {
 				onChannelCreate.accept(ch);
 			}

--- a/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
+++ b/src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java
@@ -30,10 +30,13 @@ import io.netty.channel.EventLoopGroup;
 import io.netty.channel.pool.ChannelHealthChecker;
 import io.netty.channel.pool.ChannelPool;
 import io.netty.channel.pool.ChannelPoolHandler;
+import io.netty.util.AttributeKey;
 import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
 import io.netty.util.concurrent.Promise;
 import io.netty.util.internal.PlatformDependent;
 import reactor.core.publisher.Mono;
+import reactor.ipc.netty.NettyPipeline;
 import reactor.util.Logger;
 import reactor.util.Loggers;
 
@@ -82,7 +85,8 @@ final class DefaultPoolResources implements PoolResources {
 	}
 
 	final static class Pool extends AtomicBoolean
-			implements ChannelPoolHandler, ChannelPool, ChannelHealthChecker {
+			implements ChannelPoolHandler, ChannelPool, ChannelHealthChecker,
+			           GenericFutureListener<Future<Channel>> {
 
 		final ChannelPool               pool;
 		final Consumer<? super Channel> onChannelCreate;
@@ -114,12 +118,34 @@ final class DefaultPoolResources implements PoolResources {
 
 		@Override
 		public Future<Channel> acquire() {
-			return pool.acquire();
+			return acquire(defaultGroup.next().newPromise());
 		}
 
 		@Override
 		public Future<Channel> acquire(Promise<Channel> promise) {
-			return pool.acquire(promise);
+			return pool.acquire(promise).addListener(this);
+		}
+
+		@Override
+		public void operationComplete(Future<Channel> future) throws Exception {
+			if (future.isSuccess() ){
+				Channel c = future.get();
+				activeConnections.incrementAndGet();
+				if (log.isDebugEnabled()) {
+					log.debug("Acquired {}, now {} active connections",
+							c.toString(),
+							activeConnections);
+				}
+
+
+				if(c.attr(CLOSE_HANDLER_ADDED).setIfAbsent(true) == null) {
+					if (log.isDebugEnabled()) {
+						log.debug("Registering close event to pool release: {}", c.toString());
+					}
+					c.closeFuture()
+					 .addListener(ff -> pool.release(c));
+				}
+			}
 		}
 
 		@Override
@@ -151,17 +177,11 @@ final class DefaultPoolResources implements PoolResources {
 
 		@Override
 		public void channelAcquired(Channel ch) throws Exception {
-			activeConnections.incrementAndGet();
-			if (log.isDebugEnabled()) {
-				log.debug("Acquired {}, now {} active connections",
-						ch.toString(),
-						activeConnections);
-			}
+
 		}
 
 		@Override
 		public void channelCreated(Channel ch) throws Exception {
-			activeConnections.incrementAndGet();
 			/*
 				Sometimes the Channel can be notified as created (by FixedChannelPool) but
 				it actually fails to connect and the FixedChannelPool will decrement its
@@ -217,6 +237,9 @@ final class DefaultPoolResources implements PoolResources {
 	}
 
 	static final Logger log = Loggers.getLogger(DefaultPoolResources.class);
+
+	static final AttributeKey<Boolean> CLOSE_HANDLER_ADDED = AttributeKey.valueOf
+			("closeHandlerAdded");
 
 
 	final static class SocketAddressHolder {

--- a/src/test/java/reactor/ipc/netty/resources/DefaultPoolResourcesTest.java
+++ b/src/test/java/reactor/ipc/netty/resources/DefaultPoolResourcesTest.java
@@ -15,20 +15,32 @@
  */
 package reactor.ipc.netty.resources;
 
+import java.io.IOException;
 import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.DefaultEventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
 import io.netty.channel.pool.ChannelPool;
+import io.netty.channel.socket.nio.NioSocketChannel;
 import io.netty.util.concurrent.Future;
 import io.netty.util.concurrent.Promise;
 import org.junit.Before;
 import org.junit.Test;
 import reactor.core.publisher.Mono;
+import reactor.ipc.netty.SocketUtils;
+import reactor.ipc.netty.tcp.TcpClientTests;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 
 public class DefaultPoolResourcesTest {
 
@@ -112,6 +124,70 @@ public class DefaultPoolResourcesTest {
 		poolResources.dispose();
 
 		assertThat(closed.get()).as("pool closed only once").isEqualTo(1);
+	}
+
+	@Test
+	public void fixedPoolTwoAcquire()
+			throws ExecutionException, InterruptedException, IOException {
+		final ScheduledExecutorService service = Executors.newScheduledThreadPool(2);
+		int echoServerPort = SocketUtils.findAvailableTcpPort();
+		TcpClientTests.EchoServer echoServer = new TcpClientTests.EchoServer(echoServerPort);
+
+		List<Channel> createdChannels = new ArrayList<>();
+
+		try {
+			final InetSocketAddress address = InetSocketAddress.createUnresolved("localhost", echoServerPort);
+			ChannelPool pool = PoolResources.fixed("fixedPoolTwoAcquire", 2)
+			                                .selectOrCreate(address,
+					                                () -> new Bootstrap()
+							                                .remoteAddress(address)
+							                                .channelFactory(NioSocketChannel::new)
+							                                .group(new NioEventLoopGroup(2)),
+					                                createdChannels::add,
+					                                new NioEventLoopGroup(2));
+
+			//fail a couple
+			assertThatExceptionOfType(Throwable.class)
+					.isThrownBy(pool.acquire()::get)
+					.withMessageContaining("Connection refused");
+			assertThatExceptionOfType(Throwable.class)
+					.isThrownBy(pool.acquire()::get)
+					.withMessageContaining("Connection refused");
+
+			//start the echo server
+			service.submit(echoServer);
+			Thread.sleep(100);
+
+			//acquire 2
+			final Channel channel1 = pool.acquire().get();
+			final Channel channel2 = pool.acquire().get();
+
+			//make room for 1 more
+			channel2.close().get();
+			final Channel channel3 = pool.acquire().get();
+
+			//next one will block until a previous one is released
+			long start = System.currentTimeMillis();
+			service.schedule(() -> pool.release(channel1), 500, TimeUnit.MILLISECONDS);
+			final Channel channel4 = pool.acquire().get();
+			long end = System.currentTimeMillis();
+
+			assertThat(end - start)
+					.as("channel4 acquire blocked until channel1 released")
+					.isGreaterThan(500);
+
+			pool.release(channel3).get();
+			pool.release(channel4).get();
+
+			assertThat(pool).isInstanceOf(DefaultPoolResources.Pool.class);
+			DefaultPoolResources.Pool defaultPool = (DefaultPoolResources.Pool) pool;
+			assertThat(defaultPool.activeConnections.get())
+					.as("activeConnections fully released")
+					.isZero();
+		}
+		finally {
+			echoServer.close();
+		}
 	}
 
 }

--- a/src/test/java/reactor/ipc/netty/tcp/TcpClientTests.java
+++ b/src/test/java/reactor/ipc/netty/tcp/TcpClientTests.java
@@ -32,6 +32,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Consumer;
 
+import io.netty.channel.Channel;
 import io.netty.channel.ChannelOption;
 import io.netty.handler.codec.LineBasedFrameDecoder;
 import org.assertj.core.api.Assertions;
@@ -172,7 +173,69 @@ public class TcpClientTests {
 						                                                     NettyPipeline.ReactiveBridge,
 						                                                     "codec",
 						                                                     new LineBasedFrameDecoder(
-								                                                     8 * 1024))))
+								                                                     8 * 1024)))
+				)
+				         .newHandler((in, out) ->
+					        out.sendString(Flux.range(1, messages)
+					                            .map(i -> "Hello World!" + i + "\n")
+					                            .subscribeOn(Schedulers.parallel()))
+					            .then( in.receive()
+					                     .asString()
+					                     .take(100)
+					                     .flatMapIterable(s -> Arrays.asList(s.split("\\n")))
+					                     .doOnNext(s -> {
+						                     strings.add(s);
+						                     latch.countDown();
+					                     }).then())
+				         )
+				         .block(Duration.ofSeconds(15))
+				         .onClose()
+				         .block(Duration.ofSeconds(30));
+
+		assertTrue("Expected messages not received. Received " + strings.size() + " messages: " + strings,
+				latch.await(15, TimeUnit.SECONDS));
+
+		assertEquals(messages, strings.size());
+	}
+
+	@Test
+	public void tcpClientHandlesLineFeedDataFixedPool() throws InterruptedException {
+		Consumer<Channel> channelInit = c -> c
+				.pipeline()
+				.addBefore(NettyPipeline.ReactiveBridge,
+						"codec",
+						new LineBasedFrameDecoder(8 * 1024));
+
+		tcpClientHandlesLineFeedData(opts -> opts
+				.host("localhost")
+				    .port(echoServerPort)
+				    .poolResources(PoolResources.fixed("tcpClientHandlesLineFeedDataFixedPool", 1))
+				    .afterChannelInit(channelInit)
+		);
+	}
+
+	@Test
+	public void tcpClientHandlesLineFeedDataElasticPool() throws InterruptedException {
+		Consumer<Channel> channelInit = c -> c
+				.pipeline()
+				.addBefore(NettyPipeline.ReactiveBridge,
+						"codec",
+						new LineBasedFrameDecoder(8 * 1024));
+
+		tcpClientHandlesLineFeedData(opts -> opts
+				.host("localhost")
+				    .port(echoServerPort)
+				    .poolResources(PoolResources.elastic("tcpClientHandlesLineFeedDataElasticPool"))
+				    .afterChannelInit(channelInit)
+		);
+	}
+
+	private void tcpClientHandlesLineFeedData(Consumer<? super ClientOptions.Builder> opsBuilder) throws InterruptedException {
+		final int messages = 100;
+		final CountDownLatch latch = new CountDownLatch(messages);
+		final List<String> strings = new ArrayList<>();
+
+		TcpClient.create(opsBuilder)
 				         .newHandler((in, out) ->
 					        out.sendString(Flux.range(1, messages)
 					                            .map(i -> "Hello World!" + i + "\n")
@@ -417,7 +480,7 @@ public class TcpClientTests {
 		          .isNotSameAs(client.options());
 	}
 
-	private static final class EchoServer
+	public static final class EchoServer
 			extends CountDownLatch
 			implements Runnable {
 
@@ -425,7 +488,7 @@ public class TcpClientTests {
 		private final    ServerSocketChannel server;
 		private volatile Thread              thread;
 
-		private EchoServer(int port) {
+		public EchoServer(int port) {
 			super(1);
 			this.port = port;
 			try {

--- a/src/test/java/reactor/ipc/netty/tcp/TcpClientTests.java
+++ b/src/test/java/reactor/ipc/netty/tcp/TcpClientTests.java
@@ -30,6 +30,7 @@ import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
 
 import io.netty.channel.ChannelOption;
 import io.netty.handler.codec.LineBasedFrameDecoder;
@@ -46,6 +47,7 @@ import reactor.ipc.netty.NettyPipeline;
 import reactor.ipc.netty.SocketUtils;
 import reactor.ipc.netty.channel.AbortedException;
 import reactor.ipc.netty.http.client.HttpClient;
+import reactor.ipc.netty.options.ClientOptions;
 import reactor.ipc.netty.resources.PoolResources;
 
 import static org.hamcrest.MatcherAssert.assertThat;
@@ -207,15 +209,12 @@ public class TcpClientTests {
 		      .block(Duration.ofSeconds(30));
 	}
 
-	@Test
-	public void connectionWillRetryConnectionAttemptWhenItFailsElastic()
+	private void connectionWillRetryConnectionAttemptWhenItFails(Consumer<ClientOptions.Builder> opsBuilder)
 			throws InterruptedException {
 		final CountDownLatch latch = new CountDownLatch(1);
 		final AtomicLong totalDelay = new AtomicLong();
 
-		TcpClient.create(ops -> ops.host("localhost")
-		                           .port(abortServerPort + 3)
-		                           .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 100))
+		TcpClient.create(opsBuilder)
 		         .newHandler((in, out) -> Mono.never())
 		         .retryWhen(errors -> errors.zipWith(Flux.range(1, 4), (a, b) -> b)
 		                                    .flatMap(attempt -> {
@@ -244,43 +243,24 @@ public class TcpClientTests {
 		assertThat("totalDelay was >1.6s", totalDelay.get(), greaterThanOrEqualTo(1600L));
 	}
 
+	@Test
+	public void connectionWillRetryConnectionAttemptWhenItFailsElastic()
+			throws InterruptedException {
+		connectionWillRetryConnectionAttemptWhenItFails(ops -> ops
+				.host("localhost")
+				.port(abortServerPort + 3)
+				.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 100));
+	}
+
 	//see https://github.com/reactor/reactor-netty/issues/289
 	@Test
 	public void connectionWillRetryConnectionAttemptWhenItFailsFixedChannelPool()
 			throws InterruptedException {
-		final CountDownLatch latch = new CountDownLatch(1);
-		final AtomicLong totalDelay = new AtomicLong();
-
-		TcpClient.create(ops -> ops.host("localhost")
-		                           .port(abortServerPort + 3)
-		                           .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 100)
-		                           .poolResources(PoolResources.fixed("test", 1)))
-		         .newHandler((in, out) -> Mono.never())
-		         .retryWhen(errors -> errors.zipWith(Flux.range(1, 4), (a, b) -> b)
-		                                    .flatMap(attempt -> {
-			                                    switch (attempt) {
-				                                    case 1:
-					                                    totalDelay.addAndGet(100);
-					                                    return Mono.delay(Duration
-							                                    .ofMillis(100));
-				                                    case 2:
-					                                    totalDelay.addAndGet(500);
-					                                    return Mono.delay(Duration
-							                                    .ofMillis(500));
-				                                    case 3:
-					                                    totalDelay.addAndGet(1000);
-					                                    return Mono.delay(Duration
-							                                    .ofSeconds(1));
-				                                    default:
-					                                    latch.countDown();
-					                                    return Mono.<Long>empty();
-			                                    }
-		                                    }))
-		         .subscribe(System.out::println);
-
-		latch.await(5, TimeUnit.SECONDS);
-		assertTrue("latch was counted down:" + latch.getCount(), latch.getCount() == 0);
-		assertThat("totalDelay was >1.6s", totalDelay.get(), greaterThanOrEqualTo(1600L));
+		connectionWillRetryConnectionAttemptWhenItFails(ops -> ops
+				.host("localhost")
+				.port(abortServerPort + 3)
+				.poolResources(PoolResources.fixed("test", 1))
+				.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 100));
 	}
 
 	@Test

--- a/src/test/java/reactor/ipc/netty/tcp/TcpClientTests.java
+++ b/src/test/java/reactor/ipc/netty/tcp/TcpClientTests.java
@@ -46,6 +46,7 @@ import reactor.ipc.netty.NettyPipeline;
 import reactor.ipc.netty.SocketUtils;
 import reactor.ipc.netty.channel.AbortedException;
 import reactor.ipc.netty.http.client.HttpClient;
+import reactor.ipc.netty.resources.PoolResources;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
@@ -207,7 +208,7 @@ public class TcpClientTests {
 	}
 
 	@Test
-	public void connectionWillRetryConnectionAttemptWhenItFails()
+	public void connectionWillRetryConnectionAttemptWhenItFailsElastic()
 			throws InterruptedException {
 		final CountDownLatch latch = new CountDownLatch(1);
 		final AtomicLong totalDelay = new AtomicLong();
@@ -215,6 +216,45 @@ public class TcpClientTests {
 		TcpClient.create(ops -> ops.host("localhost")
 		                           .port(abortServerPort + 3)
 		                           .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 100))
+		         .newHandler((in, out) -> Mono.never())
+		         .retryWhen(errors -> errors.zipWith(Flux.range(1, 4), (a, b) -> b)
+		                                    .flatMap(attempt -> {
+			                                    switch (attempt) {
+				                                    case 1:
+					                                    totalDelay.addAndGet(100);
+					                                    return Mono.delay(Duration
+							                                    .ofMillis(100));
+				                                    case 2:
+					                                    totalDelay.addAndGet(500);
+					                                    return Mono.delay(Duration
+							                                    .ofMillis(500));
+				                                    case 3:
+					                                    totalDelay.addAndGet(1000);
+					                                    return Mono.delay(Duration
+							                                    .ofSeconds(1));
+				                                    default:
+					                                    latch.countDown();
+					                                    return Mono.<Long>empty();
+			                                    }
+		                                    }))
+		         .subscribe(System.out::println);
+
+		latch.await(5, TimeUnit.SECONDS);
+		assertTrue("latch was counted down:" + latch.getCount(), latch.getCount() == 0);
+		assertThat("totalDelay was >1.6s", totalDelay.get(), greaterThanOrEqualTo(1600L));
+	}
+
+	//see https://github.com/reactor/reactor-netty/issues/289
+	@Test
+	public void connectionWillRetryConnectionAttemptWhenItFailsFixedChannelPool()
+			throws InterruptedException {
+		final CountDownLatch latch = new CountDownLatch(1);
+		final AtomicLong totalDelay = new AtomicLong();
+
+		TcpClient.create(ops -> ops.host("localhost")
+		                           .port(abortServerPort + 3)
+		                           .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 100)
+		                           .poolResources(PoolResources.fixed("test", 1)))
 		         .newHandler((in, out) -> Mono.never())
 		         .retryWhen(errors -> errors.zipWith(Flux.range(1, 4), (a, b) -> b)
 		                                    .flatMap(attempt -> {

